src/sql-parser/src/ast/defs/name.rs

```diff
@@ -29,12 +30,202 @@ use crate::ast::{AstInfo, QualifiedReplica};
+        }
+
+        let mut byte_length = 0;
+        let s_truncated = s
+            .chars()
+            .take_while(|c| {
+                byte_length += c.len_utf8();
+                byte_length < Self::MAX_LENGTH
+            })
+            .collect();
+
+        Ident(s_truncated)
+    }
+
+    /// Create a new [`Ident`] _without checking any of our invariants_.
+    ///
+    /// NOTE: Generally you __should not use this function__! If you're trying to create an
+    /// [`Ident`] from a `&'static str` you know is valid, use the [`ident!`] macro. For all other
+    /// use cases, see [`Ident::new`] which correctly checks our invariants.
+    ///
+    /// [`ident!`]: [`mz_sql_parser::ident`]
+    pub fn new_unchecked<S: Into<String>>(value: S) -> Self {
+        let s = value.into();
+        mz_ore::soft_assert!(s.len() <= Self::MAX_LENGTH);
+
+        Ident(s)
+    }
+
+    /// Generate a valid [`Ident`] with the provided `prefix` and `suffix`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use mz_sql_parser::ast::{Ident, IdentError};
+    ///
+    /// let good_id =
+    ///   Ident::try_generate_name("hello", "_world", |_| Ok::<_, IdentError>(true)).unwrap();
+    /// assert_eq!(good_id.as_str(), "hello_world");
+    ///
+    /// // Return invalid once.
+    /// let mut attempts = 0;
+    /// let one_failure = Ident::try_generate_name("hello", "_world", |_candidate| {
+    ///     if attempts == 0 {
+    ///         attempts += 1;
+    ///         Ok::<_, IdentError>(false)
+    ///     } else {
+    ///         Ok(true)
+    ///     }
+    /// })
+    /// .unwrap();
+    ///
+    /// // We "hello_world" was invalid, so we appended "_1".
+    /// assert_eq!(one_failure.as_str(), "hello_world_1");
+    /// ```
+    pub fn try_generate_name<P, S, F, E>(prefix: P, suffix: S, mut is_valid: F) -> Result<Self, E>
+    where
+        P: Into<String>,
+        S: Into<String>,
+        E: From<IdentError>,
+        F: FnMut(&Ident) -> Result<bool, E>,
+    {
+        const MAX_ATTEMPTS: usize = 1000;
+
+        let prefix: String = prefix.into();
+        let suffix: String = suffix.into();
+
+        // First just append the prefix and suffix.
+        let mut candidate = Ident(prefix.clone());
+        candidate.append_lossy(suffix.clone());
+        if is_valid(&candidate)? {
+            return Ok(candidate);
+        }
+
+        // Otherwise, append a number to the back.
+        for i in 1..MAX_ATTEMPTS {
+            let mut candidate = Ident(prefix.clone());
+            candidate.append_lossy(format!("{suffix}_{i}"));
+
+            if is_valid(&candidate)? {
+                return Ok(candidate);
+            }
+        }
+
+        // Couldn't find any valid name!
+        Err(E::from(IdentError::FailedToGenerate {
+            prefix,
+            suffix,
+            attempts: MAX_ATTEMPTS,
+        }))
+    }
+
+    /// Append the provided `suffix`, truncating `self` as necessary to satisfy our invariants.
```

Comment:
Maybe add that `suffix` can be truncated too.

Reply:
Added a doc test to show this